File: platform/L4-integration/reconcile/reconcile.go
// Package reconcile is the L4 federation drift-reconciliation engine (CC-SPEC-001 §10.6, §20).
//
// "Federate, never duplicate" only holds if the platform can TELL when its local copy has drifted from the
// national source of truth (APAAR, UDISE+/EMIS, PFMS). This is a faithful port of the reference engine
// (lib/federation/reconcile.ts, crosswalk verdict PORT). It compares an upstream record against the local
// record field-by-field, classifies each field, and produces an EXPLAINABLE advisory verdict —
// Reconciled / Review / Flagged. It is advisory only and never mutates anything: a human reconciler decides
// (HITL). Pure + deterministic.
package reconcile

import (
	"fmt"
	"math"
	"strings"
)

// Recommendation is the advisory verdict.
type Recommendation string

const (
	Reconciled Recommendation = "Reconciled" // local agrees with the source of truth
	Review     Recommendation = "Review"     // non-critical fields differ; reconcile to source
	Flagged    Recommendation = "Flagged"    // identity-critical drift; verify before trusting either copy
)

// ── String (field-level) reconciliation ─────────────────────────────────────

// FieldState classifies one string field.
type FieldState string

const (
	Match           FieldState = "match"
	Drift           FieldState = "drift"
	MissingUpstream FieldState = "missing-upstream"
	MissingLocal    FieldState = "missing-local"
)

// FieldComparison is one compared field.
type FieldComparison struct {
	Field    string
	Label    string
	Upstream string
	Local    string
	State    FieldState
	Critical bool // identity-critical (name, DOB) — a drift here escalates to Flagged
}

// Report is the string-reconciliation result.
type Report struct {
	Recommendation     Recommendation
	Rationale          string
	Fields             []FieldComparison
	Comparable         int
	Matches            int
	DriftCount         int
	CriticalDriftCount int
	MatchPct           int
}

func norm(v string) string { return strings.ToLower(strings.TrimSpace(v)) }

// ClassifyField classifies a string field from raw upstream/local values.
func ClassifyField(upstream, local string) FieldState {
	u, l := norm(upstream), norm(local)
	switch {
	case u == "" && l == "":
		return Match // nothing to compare → not a discrepancy
	case u == "":
		return MissingUpstream
	case l == "":
		return MissingLocal
	case u == l:
		return Match
	default:
		return Drift
	}
}

// Field is the raw input for one comparison.
type Field struct {
	Field    string
	Label    string
	Upstream string
	Local    string
	Critical bool
}

// Compare classifies each raw field and builds the report.
func Compare(rows []Field) Report {
	fields := make([]FieldComparison, len(rows))
	for i, r := range rows {
		fields[i] = FieldComparison{
			Field: r.Field, Label: r.Label, Upstream: r.Upstream, Local: r.Local,
			State: ClassifyField(r.Upstream, r.Local), Critical: r.Critical,
		}
	}
	return buildReport(fields)
}

func bothEmpty(f FieldComparison) bool { return norm(f.Upstream) == "" && norm(f.Local) == "" }

func buildReport(fields []FieldComparison) Report {
	comparable, matches := 0, 0
	var drifts, criticalDrifts []FieldComparison
	for _, f := range fields {
		if !bothEmpty(f) {
			comparable++
			if f.State == Match {
				matches++
			}
		}
		if f.State != Match {
			drifts = append(drifts, f)
			if f.Critical {
				criticalDrifts = append(criticalDrifts, f)
			}
		}
	}
	matchPct := 100
	if comparable > 0 {
		matchPct = int(math.Round(float64(matches) / float64(comparable) * 100))
	}

	var rec Recommendation
	var rationale string
	switch {
	case len(criticalDrifts) > 0:
		rec = Flagged
		rationale = fmt.Sprintf("Identity-critical drift on %s — verify before trusting either copy.", labels(criticalDrifts))
	case len(drifts) > 0:
		rec = Review
		rationale = fmt.Sprintf("%d non-critical field(s) differ (%s); reconcile the local record to the source of truth.", len(drifts), labels(drifts))
	default:
		rec = Reconciled
		if comparable == 0 {
			rationale = "No overlapping fields to compare."
		} else {
			rationale = "Local record agrees with the source of truth on every compared field."
		}
	}
	return Report{
		Recommendation: rec, Rationale: rationale, Fields: fields,
		Comparable: comparable, Matches: matches, DriftCount: len(drifts),
		CriticalDriftCount: len(criticalDrifts), MatchPct: matchPct,
	}
}

func labels(fs []FieldComparison) string {
	out := make([]string, len(fs))
	for i, f := range fs {
		out[i] = f.Label
	}
	return strings.Join(out, ", ")
}

// ── Numeric (count/money) reconciliation — tolerance-aware ──────────────────

// NumericState classifies one numeric field against a tolerance.
type NumericState string

const (
	NumMatch        NumericState = "match"
	MinorDrift      NumericState = "minor-drift" // within tolerance → counts as agreement
	NumDrift        NumericState = "drift"       // beyond tolerance → a real discrepancy
	NumMissingUp    NumericState = "missing-upstream"
	NumMissingLocal NumericState = "missing-local"
)

// DefaultTolerancePct: count deltas at/under this percentage are "minor", not a discrepancy.
const DefaultTolerancePct = 2.0

// DefaultMoneyTolerancePct: scheme funds reconcile tighter than head-counts.
const DefaultMoneyTolerancePct = 1.0

// NumericComparison is one numeric field; Upstream/Local nil means that side keeps no master.
type NumericComparison struct {
	Field    string
	Label    string
	Upstream *float64
	Local    *float64
	Delta    float64 // local − upstream (0 when a side is missing)
	PctDelta int     // |delta|/upstream as a whole percent
	State    NumericState
	Critical bool
}

// NumericReport is the numeric-reconciliation result.
type NumericReport struct {
	Recommendation     Recommendation
	Rationale          string
	Fields             []NumericComparison
	Comparable         int
	Matches            int
	DriftCount         int
	CriticalDriftCount int
	TolerancePct       float64
}

// ClassifyNumeric grades a numeric field against a tolerance.
func ClassifyNumeric(upstream, local *float64, tolerancePct float64) NumericState {
	switch {
	case upstream == nil && local == nil:
		return NumMatch
	case upstream == nil:
		return NumMissingUp
	case local == nil:
		return NumMissingLocal
	}
	delta := math.Abs(*local - *upstream)
	if delta == 0 {
		return NumMatch
	}
	pct := 100.0
	if *upstream != 0 {
		pct = delta / *upstream * 100
	}
	if pct <= tolerancePct {
		return MinorDrift
	}
	return NumDrift
}

// NumField is raw numeric input; nil pointers denote a missing master on that side.
type NumField struct {
	Field    string
	Label    string
	Upstream *float64
	Local    *float64
	Critical bool
}

// F is a helper to take the address of a float literal.
func F(v float64) *float64 { return &v }

// CompareNumeric classifies each numeric field and builds the report.
func CompareNumeric(rows []NumField, tolerancePct float64) NumericReport {
	fields := make([]NumericComparison, len(rows))
	for i, r := range rows {
		state := ClassifyNumeric(r.Upstream, r.Local, tolerancePct)
		both := r.Upstream != nil && r.Local != nil
		var delta float64
		pct := 0
		if both {
			delta = *r.Local - *r.Upstream
			if *r.Upstream != 0 {
				pct = int(math.Round(math.Abs(delta) / *r.Upstream * 100))
			} else {
				pct = 100
			}
		}
		fields[i] = NumericComparison{
			Field: r.Field, Label: r.Label, Upstream: r.Upstream, Local: r.Local,
			Delta: delta, PctDelta: pct, State: state, Critical: r.Critical,
		}
	}
	return buildNumericReport(fields, tolerancePct)
}

func buildNumericReport(fields []NumericComparison, tolerancePct float64) NumericReport {
	comparable, matches, driftCount := 0, 0, 0
	var realDrifts, criticalDrifts []NumericComparison
	for _, f := range fields {
		present := f.Upstream != nil || f.Local != nil
		if present {
			comparable++
			if f.State == NumMatch || f.State == MinorDrift {
				matches++
			}
		}
		if f.State == NumDrift || f.State == NumMissingUp || f.State == NumMissingLocal {
			driftCount++
		}
		if f.State == NumDrift {
			realDrifts = append(realDrifts, f)
			if f.Critical {
				criticalDrifts = append(criticalDrifts, f)
			}
		}
	}

	var rec Recommendation
	var rationale string
	switch {
	case len(criticalDrifts) > 0:
		rec = Flagged
		rationale = fmt.Sprintf("Count drift beyond %g%% tolerance on %s — investigate the local figure against the state master.", tolerancePct, numLabels(criticalDrifts))
	case len(realDrifts) > 0:
		rec = Review
		rationale = fmt.Sprintf("%d count(s) differ beyond tolerance (%s); reconcile with the state master.", len(realDrifts), numLabels(realDrifts))
	default:
		rec = Reconciled
		if comparable == 0 {
			rationale = "No overlapping counts to compare."
		} else {
			rationale = fmt.Sprintf("Local figures agree with the state master within the %g%% sync tolerance.", tolerancePct)
		}
	}
	return NumericReport{
		Recommendation: rec, Rationale: rationale, Fields: fields,
		Comparable: comparable, Matches: matches, DriftCount: driftCount,
		CriticalDriftCount: len(criticalDrifts), TolerancePct: tolerancePct,
	}
}

func numLabels(fs []NumericComparison) string {
	out := make([]string, len(fs))
	for i, f := range fs {
		out[i] = f.Label
	}
	return strings.Join(out, ", ")
}

// ── Concrete federation comparators ─────────────────────────────────────────

// MapJourneyToStatus maps an APAAR journeyStatus to the local student status.
func MapJourneyToStatus(journey string) string {
	switch norm(journey) {
	case "enrolled":
		return "Enrolled"
	case "transferred":
		return "Transferred"
	case "alumni":
		return "Graduated"
	case "dropout":
		return "Dropped"
	default:
		return ""
	}
}

// ApaarRecord / StudentRecord are the minimal shapes the APAAR comparator needs.
type ApaarRecord struct {
	ApaarID, Name, DateOfBirth, Gender, Category, JourneyStatus string
}
type StudentRecord struct {
	ApaarID, Name, DOB, Gender, Category, Status string
}

// CompareApaarToStudent reconciles an APAAR record (source of truth) against the local student master.
// Identity-critical: APAAR id, name, date of birth.
func CompareApaarToStudent(u ApaarRecord, l StudentRecord) Report {
	return Compare([]Field{
		{Field: "apaarId", Label: "APAAR id", Upstream: u.ApaarID, Local: l.ApaarID, Critical: true},
		{Field: "name", Label: "Name", Upstream: u.Name, Local: l.Name, Critical: true},
		{Field: "dob", Label: "Date of birth", Upstream: u.DateOfBirth, Local: l.DOB, Critical: true},
		{Field: "gender", Label: "Gender", Upstream: u.Gender, Local: l.Gender},
		{Field: "category", Label: "Category", Upstream: u.Category, Local: l.Category},
		{Field: "status", Label: "Journey / status", Upstream: MapJourneyToStatus(u.JourneyStatus), Local: l.Status},
	})
}

// PfmsExpenditure / FundLedger are the minimal shapes the PFMS comparator needs (in paisa/rupees).
type PfmsExpenditure struct{ Allocated, Released, Utilised float64 }
type FundLedger struct{ Allocated, Released, Utilised float64 }

// CompareFundFlowToPfms reconciles PFMS fund-flow figures (source of truth) against the local scheme ledger.
// Every money field is identity-critical; tolerance is tight (drift = potential leakage/mis-posting).
func CompareFundFlowToPfms(p PfmsExpenditure, l *FundLedger) NumericReport {
	var al, rl, ut *float64
	if l != nil {
		al, rl, ut = F(l.Allocated), F(l.Released), F(l.Utilised)
	}
	return CompareNumeric([]NumField{
		{Field: "allocated", Label: "Allocated", Upstream: F(p.Allocated), Local: al, Critical: true},
		{Field: "released", Label: "Released", Upstream: F(p.Released), Local: rl, Critical: true},
		{Field: "utilised", Label: "Utilised", Upstream: F(p.Utilised), Local: ut, Critical: true},
	}, DefaultMoneyTolerancePct)
}

File: lib/federation/reconcile.ts
// VASA-EOS(SE) — Federation reconciliation: field-level drift detection (L4, AI-native, advisory).
//
// "Federate, never duplicate" only works if the platform can TELL when its local copy has drifted
// from the national source of truth. This pure engine compares an upstream registry record against
// the matching local record field by field, classifies each field (match / drift / missing on one
// side) and produces an explainable advisory recommendation — Reconciled, Review or Flagged.
//
// It is advisory only: a human reconciler decides (HITL). The engine never mutates anything; it just
// makes the discrepancy visible and reasoned, the way the AI engines do. Deterministic + client-safe.

import type { ApaarRecord, EmisSchoolData } from "@/lib/integrations/types"
import type { StudentRecord } from "@/lib/students"
import type { EnrolmentRecord } from "@/lib/enrolment/store"

export type FieldState = "match" | "drift" | "missing-upstream" | "missing-local"

export interface FieldComparison {
  field: string
  label: string
  upstream: string
  local: string
  state: FieldState
  /** Identity-critical fields (name, date of birth) escalate a drift to a Flag. */
  critical: boolean
}

export const RECONCILE_RECOMMENDATIONS = ["Reconciled", "Review", "Flagged"] as const
export type ReconcileRecommendation = (typeof RECONCILE_RECOMMENDATIONS)[number]

export interface ReconcileReport {
  recommendation: ReconcileRecommendation
  rationale: string
  fields: FieldComparison[]
  /** Fields with at least one side present (the basis for matchPct). */
  comparable: number
  matches: number
  driftCount: number
  criticalDriftCount: number
  /** Whole-number percentage of comparable fields that agree. */
  matchPct: number
}

const norm = (v: string | undefined | null): string => (v ?? "").trim().toLowerCase()

/** Classify a single field given the (raw) upstream and local values. */
export function classifyField(upstream: string, local: string): FieldState {
  const u = norm(upstream)
  const l = norm(local)
  if (!u && !l) return "match" // nothing to compare → not a discrepancy
  if (!u) return "missing-upstream"
  if (!l) return "missing-local"
  return u === l ? "match" : "drift"
}

/** APAAR journeyStatus ↔ local StudentStatus, normalised so "enrolled" and "Enrolled" agree. */
export function mapJourneyToStatus(journey: string | undefined): string {
  switch (norm(journey)) {
    case "enrolled":
      return "Enrolled"
    case "transferred":
      return "Transferred"
    case "alumni":
      return "Graduated"
    case "dropout":
      return "Dropped"
    default:
      return ""
  }
}

/** Build the report (recommendation + totals) from a set of compared fields. Pure. */
export function buildReport(fields: FieldComparison[]): ReconcileReport {
  const comparable = fields.filter((f) => !(norm(f.upstream) === "" && norm(f.local) === "")).length
  const matches = fields.filter((f) => f.state === "match" && !(norm(f.upstream) === "" && norm(f.local) === "")).length
  const drifts = fields.filter((f) => f.state !== "match")
  const driftCount = drifts.length
  const criticalDriftCount = drifts.filter((f) => f.critical).length
  const matchPct = comparable === 0 ? 100 : Math.round((matches / comparable) * 100)

  let recommendation: ReconcileRecommendation
  let rationale: string
  if (criticalDriftCount > 0) {
    recommendation = "Flagged"
    rationale = `Identity-critical drift on ${drifts.filter((f) => f.critical).map((f) => f.label).join(", ")} — verify before trusting either copy.`
  } else if (driftCount > 0) {
    recommendation = "Review"
    rationale = `${driftCount} non-critical field(s) differ (${drifts.map((f) => f.label).join(", ")}); reconcile the local record to the source of truth.`
  } else {
    recommendation = "Reconciled"
    rationale = comparable === 0 ? "No overlapping fields to compare." : "Local record agrees with the source of truth on every compared field."
  }
  return { recommendation, rationale, fields, comparable, matches, driftCount, criticalDriftCount, matchPct }
}

/**
 * Compare an upstream APAAR record (source of truth) against the local student master record.
 * Identity-critical: name, date of birth. Advisory only — a human decides the reconciliation.
 */
export function compareApaarToStudent(upstream: ApaarRecord, local: StudentRecord): ReconcileReport {
  const rows: Array<{ field: string; label: string; u: string; l: string; critical: boolean }> = [
    { field: "apaarId", label: "APAAR id", u: upstream.apaarId ?? "", l: local.apaarId ?? "", critical: true },
    { field: "name", label: "Name", u: upstream.name ?? "", l: local.name ?? "", critical: true },
    { field: "dob", label: "Date of birth", u: upstream.dateOfBirth ?? "", l: local.dob ?? "", critical: true },
    { field: "gender", label: "Gender", u: upstream.gender ?? "", l: local.gender ?? "", critical: false },
    { field: "category", label: "Category", u: upstream.category ?? "", l: local.category ?? "", critical: false },
    { field: "status", label: "Journey / status", u: mapJourneyToStatus(upstream.journeyStatus), l: local.status ?? "", critical: false },
  ]
  const fields: FieldComparison[] = rows.map((r) => ({
    field: r.field,
    label: r.label,
    upstream: r.u,
    local: r.l,
    state: classifyField(r.u, r.l),
    critical: r.critical,
  }))
  return buildReport(fields)
}

// ── Numeric reconciliation (counts) — tolerance-aware, for EMIS/UDISE+ master-data ──────────────
//
// String equality is wrong for counts: a state EMIS snapshot lags the school roll by a few records,
// so a small percentage delta is "within tolerance", not a discrepancy. This comparator grades a
// numeric field as match / minor-drift / drift (or missing on either side) against a tolerance.

export type NumericState = "match" | "minor-drift" | "drift" | "missing-upstream" | "missing-local"

export interface NumericComparison {
  field: string
  label: string
  /** null = the side keeps no master for this field (→ missing). */
  upstream: number | null
  local: number | null
  /** local − upstream (0 when a side is missing). */
  delta: number
  /** |delta| / upstream as a whole-number percentage (0 when upstream is missing/zero). */
  pctDelta: number
  state: NumericState
  critical: boolean
}

export interface NumericReport {
  recommendation: ReconcileRecommendation
  rationale: string
  fields: NumericComparison[]
  comparable: number
  matches: number
  driftCount: number
  criticalDriftCount: number
  /** Tolerance (%) below which a non-zero delta is graded a minor-drift, not a drift. */
  tolerancePct: number
}

/** Default sync tolerance: deltas at or under this percentage are "minor", not a real discrepancy. */
export const DEFAULT_TOLERANCE_PCT = 2

export function classifyNumeric(upstream: number | null, local: number | null, tolerancePct = DEFAULT_TOLERANCE_PCT): NumericState {
  if (upstream === null && local === null) return "match"
  if (upstream === null) return "missing-upstream"
  if (local === null) return "missing-local"
  const delta = Math.abs(local - upstream)
  if (delta === 0) return "match"
  const pct = upstream === 0 ? 100 : (delta / upstream) * 100
  return pct <= tolerancePct ? "minor-drift" : "drift"
}

function numericField(field: string, label: string, upstream: number | null, local: number | null, critical: boolean, tolerancePct: number): NumericComparison {
  const state = classifyNumeric(upstream, local, tolerancePct)
  const both = upstream !== null && local !== null
  const delta = both ? local - upstream : 0
  const pctDelta = both && upstream !== 0 ? Math.round((Math.abs(delta) / upstream) * 100) : both && upstream === 0 ? 100 : 0
  return { field, label, upstream, local, delta, pctDelta, state, critical }
}

/** Build a numeric report (recommendation + totals). A "drift" (beyond tolerance) is a real discrepancy; a "minor-drift" is within tolerance and counts as agreement. Pure. */
export function buildNumericReport(fields: NumericComparison[], tolerancePct = DEFAULT_TOLERANCE_PCT): NumericReport {
  const comparable = fields.filter((f) => f.upstream !== null || f.local !== null).length
  const agree = fields.filter((f) => (f.state === "match" || f.state === "minor-drift") && (f.upstream !== null || f.local !== null))
  const matches = agree.length
  const drifts = fields.filter((f) => f.state === "drift" || f.state === "missing-upstream" || f.state === "missing-local")
  const realDrifts = fields.filter((f) => f.state === "drift")
  const criticalDriftCount = realDrifts.filter((f) => f.critical).length

  let recommendation: ReconcileRecommendation
  let rationale: string
  if (criticalDriftCount > 0) {
    recommendation = "Flagged"
    rationale = `Count drift beyond ${tolerancePct}% tolerance on ${realDrifts.filter((f) => f.critical).map((f) => f.label).join(", ")} — investigate the local roll against the state master.`
  } else if (realDrifts.length > 0) {
    recommendation = "Review"
    rationale = `${realDrifts.length} count(s) differ beyond tolerance (${realDrifts.map((f) => f.label).join(", ")}); reconcile with the state master.`
  } else {
    recommendation = "Reconciled"
    rationale = comparable === 0 ? "No overlapping counts to compare." : `Local counts agree with the state master within the ${tolerancePct}% sync tolerance.`
  }
  return { recommendation, rationale, fields, comparable, matches, driftCount: drifts.length, criticalDriftCount, tolerancePct }
}

/**
 * Reconcile the state EMIS master-data snapshot (counts) against the local school roll. Only the
 * on-roll student count has a local master (the enrolment snapshot); teachers/classrooms are shown
 * as upstream-only context (missing-local), honestly, rather than inventing a local figure.
 */
export function compareEmisToEnrolment(emis: EmisSchoolData, local: EnrolmentRecord | null, tolerancePct = DEFAULT_TOLERANCE_PCT): NumericReport {
  const fields: NumericComparison[] = [
    numericField("students", "Students on roll", emis.students, local ? local.total : null, true, tolerancePct),
    numericField("teachers", "Teachers", emis.teachers, null, false, tolerancePct),
    numericField("classrooms", "Classrooms", emis.classrooms, null, false, tolerancePct),
  ]
  return buildNumericReport(fields, tolerancePct)
}
